pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go

```diff
@@ -968,6 +1000,75 @@ func KeyV1(workerPoolName string, kubernetesVersion *semver.Version, criConfig *
 	return fmt.Sprintf("gardener-node-agent-%s-%s", workerPoolName, utils.ComputeSHA256Hex([]byte(kubernetesMajorMinorVersion + string(criName)))[:5])
 }
 
+// KeyV2 returns the key that can be used as secret name based on the provided worker name, Kubernetes version and CRI
+// configuration.
+func KeyV2(
+	kubernetesVersion *semver.Version,
+	credentialsRotation *gardencorev1beta1.ShootCredentialsRotation,
+	worker *gardencorev1beta1.Worker,
+	nodeLocalDNSEnabled bool,
+	kubeletConfiguration *gardencorev1beta1.KubeletConfig,
+) string {
+	if kubernetesVersion == nil {
+		return ""
+	}
+
+	kubernetesMajorMinorVersion := fmt.Sprintf("%d.%d", kubernetesVersion.Major(), kubernetesVersion.Minor())
+
+	data := []string{
+		kubernetesMajorMinorVersion,
+		worker.Machine.Type,
+		worker.Machine.Image.Name + *worker.Machine.Image.Version,
+	}
+
+	if worker.Volume != nil {
+		data = append(data, worker.Volume.VolumeSize)
+		if worker.Volume.Type != nil {
+			data = append(data, *worker.Volume.Type)
+		}
+	}
+
+	if worker.CRI != nil {
+		data = append(data, string(worker.CRI.Name))
+	}
+
+	if credentialsRotation != nil {
+		if credentialsRotation.CertificateAuthorities != nil && credentialsRotation.CertificateAuthorities.LastInitiationTime != nil {
+			data = append(data, credentialsRotation.CertificateAuthorities.LastInitiationTime.Time.String())
+		}
+		if credentialsRotation.ServiceAccountKey != nil && credentialsRotation.ServiceAccountKey.LastInitiationTime != nil {
+			data = append(data, credentialsRotation.ServiceAccountKey.LastInitiationTime.Time.String())
+		}
+	}
+
+	if nodeLocalDNSEnabled {
+		data = append(data, "node-local-dns")
+	}
+
+	if kubeletConfiguration != nil {
+		if kubeReserved := kubeletConfiguration.KubeReserved; kubeReserved != nil {
+			data = append(data, fmt.Sprintf("%s-%s-%s-%s", kubeReserved.CPU, kubeReserved.Memory, kubeReserved.PID, kubeReserved.EphemeralStorage))
+		}
+		if eviction := kubeletConfiguration.EvictionHard; eviction != nil {
+			parts := make([]string, 0, 5)
+			for _, value := range []*string{eviction.ImageFSAvailable, eviction.ImageFSInodesFree, eviction.MemoryAvailable, eviction.NodeFSAvailable, eviction.NodeFSInodesFree} {
+				parts = append(parts, ptr.Deref(value, ""))
+			}
+			data = append(data, strings.Join(parts, "-"))
+		}
+		if policy := kubeletConfiguration.CPUManagerPolicy; policy != nil {
+			data = append(data, *policy)
+		}
+	}
+
+	var result string
+	for _, v := range data {
+		result += utils.ComputeSHA256Hex([]byte(v))
+	}
+
+	return fmt.Sprintf("gardener-node-agent-%s", utils.ComputeSHA256Hex([]byte(result))[:16])
```

Comment:
Why does the name no longer contain the worker pool name? In local tests I could prevent a shoot from reconciling when having two worker pools with the same settings leading to the same hash and hence the same secret name. This would not have happened if the worker pool name would be included here as well.

/hold as it might break clusters.

Reply:
removing the name was unintentional, see https://github.com/gardener/gardener/pull/9865#discussion_r1629802706

Reply:
I've fixed my fix from yesterday to just include the worker pool name as part of the secret as in KeyV1 (and as intended in https://github.com/gardener/gardener/issues/9699 ), arghh.

Reply:
I think I've found why the workerPool.Name was missing. The idea in https://github.com/gardener/gardener/issues/9699 was to remove the worker image name, which I've apparently mixed up with the workerImage.Name as both are called "local" in the local setup.

Reply:
/hold cancel